### ISSUE TITLE
skinning weight predict wrong

### ISSUE THREAD

**Issue by TenjoinWai** [VAST-AI-Research/UniRig#20]
i follow the steps and i get giraffe_skin.fbx. but when i input it into blender and pose it , i found the weight is not correct. i have no idea which step has problem.

![Image](https://github.com/user-attachments/assets/32194a8e-43c9-43cd-9aee-398c2ea27050)

**Comment by czpcf** [comment]
Could you please provide with the commands? @TenjoinWai

**Comment by TenjoinWai** [comment]
i use exact your commands in RM. here is my env and result. 
ENV:
`addict                    2.4.0                    pypi_0    pypi`
`aiohappyeyeballs          2.6.1                    pypi_0    pypi`
`aiohttp                   3.11.18                  pypi_0    pypi`
`aiosignal                 1.3.2                    pypi_0    pypi`
`antlr4-python3-runtime    4.9.3                    pypi_0    pypi`
`asttokens                 3.0.0                    pypi_0    pypi`
`attrs                     25.3.0                   pypi_0    pypi`
`blinker                   1.9.0                    pypi_0    pypi`
`bpy                       4.2.0                    pypi_0    pypi`
`bzip2                     1.0.8                h2bbff1b_6    defaults`
`ca-certificates           2025.2.25            haa95532_0    defaults`
`ccimport                  0.4.4                    pypi_0    pypi    `
`certifi                   2025.4.26                pypi_0    pypi    `
`charset-normalizer        3.4.2                    pypi_0    pypi    `
`click                     8.2.0                    pypi_0    pypi    `
`colorama                  0.4.6                    pypi_0    pypi    `
`comm                      0.2.2                    pypi_0    pypi    `
`configargparse            1.7                      pypi_0    pypi    `
`cumm-cu120                0.4.11                   pypi_0    pypi    `
`cython                    3.1.0                    pypi_0    pypi    `
`dash                      3.0.4                    pypi_0    pypi    `
`decorator                 5.2.1                    pypi_0    pypi    `
`einops                    0.8.1                    pypi_0    pypi    `
`executing                 2.2.0                    pypi_0    pypi    `
`fast-simplification       0.1.11                   pypi_0    pypi    `
`fastjsonschema            2.21.1                   pypi_0    pypi    `
`filelock                  3.13.1                   pypi_0    pypi    `
`fire                      0.7.0                    pypi_0    pypi    `
`flash-attn                2.7.4.post1              pypi_0    pypi    `
`flask                     3.0.3                    pypi_0    pypi    `
`freetype-py               2.5.1                    pypi_0    pypi    `
`frozenlist                1.6.0                    pypi_0    pypi    `
`fsspec                    2024.6.1                 pypi_0    pypi    `
`huggingface-hub           0.31.2                   pypi_0    pypi    `
`idna                      3.10                     pypi_0    pypi    `
`imageio                   2.37.0                   pypi_0    pypi    `
`importlib-metadata        8.7.0                    pypi_0    pypi    `
`ipython                   9.2.0                    pypi_0    pypi    `
`ipython-pygments-lexers   1.1.1                    pypi_0    pypi    `
`ipywidgets                8.1.7                    pypi_0    pypi    `
`itsdangerous              2.2.0                    pypi_0    pypi    `
`jedi                      0.19.2                   pypi_0    pypi    `
`jinja2                    3.1.4                    pypi_0    pypi    `
`jsonschema                4.23.0                   pypi_0    pypi    `
`jsonschema-specifications 2025.4.1                 pypi_0    pypi    `
`jupyter-core              5.7.2                    pypi_0    pypi    `
`jupyterlab-widgets        3.0.15                   pypi_0    pypi    `
`lark                      1.2.2                    pypi_0    pypi    `
`libffi                    3.4.4                hd77b12b_1    defaults`
`lightning                 2.5.1.post0              pypi_0    pypi    `
`lightning-utilities       0.14.3                   pypi_0    pypi    `
`markupsafe                2.1.5                    pypi_0    pypi    `
`matplotlib-inline         0.1.7                    pypi_0    pypi    `
`mpmath                    1.3.0                    pypi_0    pypi    `
`multidict                 6.4.3                    pypi_0    pypi    `
`narwhals                  1.39.0                   pypi_0    pypi    `
`nbformat                  5.10.4                   pypi_0    pypi    `
`nest-asyncio              1.6.0                    pypi_0    pypi    `
`networkx                  3.3                      pypi_0    pypi    `
`ninja                     1.11.1.4                 pypi_0    pypi    `
`numpy                     1.26.4                   pypi_0    pypi    `
`omegaconf                 2.3.0                    pypi_0    pypi    `
`open3d                    0.19.0                   pypi_0    pypi    `
`openssl                   3.0.16               h3f729d1_0    defaults`
`packaging                 24.2                     pypi_0    pypi    `
`parso                     0.8.4                    pypi_0    pypi    `
`pccm                      0.4.16                   pypi_0    pypi    `
`pillow                    11.0.0                   pypi_0    pypi    `
`pip                       25.1               pyhc872135_2    defaults`
`platformdirs              4.3.8                    pypi_0    pypi    `
`plotly                    6.0.1                    pypi_0    pypi    `
`portalocker               3.1.1                    pypi_0    pypi    `
`prompt-toolkit            3.0.51                   pypi_0    pypi    `
`propcache                 0.3.1                    pypi_0    pypi    `
`psutil                    7.0.0                    pypi_0    pypi    `
`pure-eval                 0.2.3                    pypi_0    pypi    `
`pybind11                  2.13.6                   pypi_0    pypi    `
`pyglet                    2.1.6                    pypi_0    pypi    `
`pygments                  2.19.1                   pypi_0    pypi    `
`pyopengl                  3.1.0                    pypi_0    pypi    `
`pyrender                  0.1.45                   pypi_0    pypi    `
`python                    3.11.11              h4607a30_0    defaults`
`python-box                7.3.2                    pypi_0    pypi    `
`pytorch-lightning         2.5.1.post0              pypi_0    pypi    `
`pywin32                   310                      pypi_0    pypi    `
`pyyaml                    6.0.2                    pypi_0    pypi    `
`referencing               0.36.2                   pypi_0    pypi    `
`regex                     2024.11.6                pypi_0    pypi    `
`requests                  2.32.3                   pypi_0    pypi    `
`retrying                  1.3.4                    pypi_0    pypi    `
`rpds-py                   0.24.0                   pypi_0    pypi    `
`safetensors               0.5.3                    pypi_0    pypi    `
`scipy                     1.15.3                   pypi_0    pypi    `
`setuptools                78.1.1          py311haa95532_0    defaults`
`six                       1.17.0                   pypi_0    pypi    `
`spconv-cu120              2.3.6                    pypi_0    pypi    `
`sqlite                    3.45.3               h2bbff1b_0    defaults`
`stack-data                0.6.3                    pypi_0    pypi    `
`sympy                     1.13.3                   pypi_0    pypi    `
`termcolor                 3.1.0                    pypi_0    pypi    `
`timm                      1.0.15                   pypi_0    pypi    `
`tk                        8.6.14               h0416ee5_0    defaults`
`tokenizers                0.21.1                   pypi_0    pypi    `
`torch                     2.7.0+cu128              pypi_0    pypi    `
`torch-cluster             1.6.3                    pypi_0    pypi    `
`torch-scatter             2.1.2                    pypi_0    pypi    `
`torchmetrics              1.7.1                    pypi_0    pypi    `
`torchvision               0.22.0+cu128             pypi_0    pypi    `
`tqdm                      4.67.1                   pypi_0    pypi    `
`traitlets                 5.14.3                   pypi_0    pypi    `
`transformers              4.51.3                   pypi_0    pypi    `
`trimesh                   4.6.9                    pypi_0    pypi    `
`triton-windows            3.3.0.post19             pypi_0    pypi    `
`typing-extensions         4.12.2                   pypi_0    pypi    `
`tzdata                    2025b                h04d1e81_0    defaults`
`urllib3                   2.4.0                    pypi_0    pypi    `
`vc                        14.42                haa95532_5    defaults`
`vs2015_runtime            14.42.34433          hbfb602d_5    defaults`
`wcwidth                   0.2.13                   pypi_0    pypi    `
`werkzeug                  3.0.6                    pypi_0    pypi    `
`wheel                     0.45.1          py311haa95532_0    defaults`
`widgetsnbextension        4.0.14                   pypi_0    pypi    `
`xz                        5.6.4                h4754444_1    defaults`
`yarl                      1.20.0                   pypi_0    pypi    `
`zipp                      3.21.0                   pypi_0    pypi    `
`zlib                      1.2.13               h8cc25b3_1    defaults`
`zstandard                 0.23.0                   pypi_0    pypi    `

CMD RESULT:
$ bash launch/inference/generate_skin.sh --input examples/skeleton/giraffe.fbx --output results/giraffe_skin.fbx
  0%|                                                    | 0/1 [00:00<?, ?it/s]Info: Deleted 1 data-block(s)
Now processing examples/skeleton/giraffe.fbx...
Now processing examples/skeleton/giraffe.fbx...
FBX version: 7400
save to: tmp\examples/skeleton/giraffe
100%|████████████████████████████████████████████| 1/1 [00:01<00:00,  1.98s/it]
1 models processed
done
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:47: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  _TORCH_CUSTOM_FWD = amp.custom_fwd(cast_inputs=torch.float16)
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:96: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:162: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:242: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:331: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:368: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:388: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\spconv\pytorch\functional.py:411: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  @_TORCH_CUSTOM_BWD
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\timm\models\layers\__init__.py:48: FutureWarning: Importing from timm.models.layers is deprecated, please import via timm.layers
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.layers", FutureWarning)
Seed set to 12345
load task config: configs/task/quick_inference_unirig_skin.yaml
load data config: configs/data\quick_inference.yaml
load transform config: configs/transform\inference_skin_transform.yaml
load model config: configs/model\unirig_skin.yaml
WARNING: use BatchNorm in ptv3obj !!!
load system config: configs/system\skin.yaml
Using bfloat16 Automatic Mixed Precision (AMP)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\lightning\pytorch\trainer\connectors\logger_connector\logger_connector.py:76: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `lightning.pytorch` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
Restoring states from the checkpoint path at C:\Users\cn200\.cache\huggingface\hub\models--VAST-AI--UniRig\snapshots\e53ad6237db5cea261727c5d50b67d140a4b9571\skin\articulation-xl\model.ckpt
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\lightning\pytorch\trainer\call.py:282: Be aware that when using `ckpt_path`, callbacks used to create the checkpoint need to be provided during `Trainer` instantiation. Please add the following callbacks: ["ModelCheckpoint{'monitor': 'val_loss_sum', 'mode': 'min', 'every_n_train_steps': 0, 'every_n_epochs': 1, 'train_time_interval': None}"].
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Loaded model weights from the checkpoint at C:\Users\cn200\.cache\huggingface\hub\models--VAST-AI--UniRig\snapshots\e53ad6237db5cea261727c5d50b67d140a4b9571\skin\articulation-xl\model.ckpt
C:\Users\cn200\anaconda3\envs\UniRig\Lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:425: The 'predict_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=15` in the `DataLoader` to improve performance.
Predicting DataLoader 0:   0%|                           | 0/1 [00:00<?, ?it/s]C:\Users\cn200\anaconda3\envs\UniRig\Lib\contextlib.py:105: FutureWarning: `torch.backends.cuda.sdp_kernel()` is deprecated. In the future, this context manager will be removed. Please see `torch.nn.attention.sdpa_kernel()` for the new context manager, with updated signature.
  self.gen = func(*args, **kwds)
E:\PcharmProjects\UniRig\src\model\unirig_skin.py:359: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  'offset': torch.tensor(batch['offset']),
E:\PcharmProjects\UniRig\src\model\unirig_skin.py:423: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  skin_pred[i, :, :num_bones[i]] = F.softmax(pred)

FBX export starting... 'results/giraffe_skin.fbx'
(bpy.data.armatures['Armature'], 'POSE')
export finished in 0.0938 sec.
Predicting DataLoader 0: 100%|███████████████████| 1/1 [00:17<00:00,  0.06it/s]
done

@czpcf

**Comment by Amble12345** [comment]
> I had the same problem running in a Windows environment. In my case, I solved the problem by building flash-attention from source code, creating a whl file and installing it.
> 
> I don't know why, but the whl file I downloaded from the link caused the same problem. The whl file I built myself did not cause the problem. https://huggingface.co/lldacing/flash-attention-windows-wheel/tree/main

Thanks for your solution! It helps a lot
